File: app/frequency.py
from app.dependencies import get_data
from app.utils import to_dict

# getting count data by usage data
def get_frequency():
  print('Getting frequency data...')
  data = get_data()
  result = {}
  for node in data:
    for dependency in data[node].connections:
      if dependency in result:
        result[dependency]['count'] += 1
      else:
        result[dependency] = {
          'count': 1,
          'data': to_dict(data[dependency])
        }
  return result 

File: src/client/js/workers/frequency.js
onmessage = function(e) {
  const { dependencies } = e.data

  // name@version: count
  const depsCount = {} 

  Object.values(dependencies).flat().forEach(item => {
    item.connections.forEach(node => {
      const namever = node.name + '@' + node.version
      if(!depsCount[namever]) {
        depsCount[namever] = 1
        return
      }
      depsCount[namever] += 1
    })
  })

  // {count: number, data: node}[]
  const result = Object.entries(depsCount).map(item => {
    const [namever, count] = item
    const separator = namever.lastIndexOf('@')
    const name = namever.slice(0, separator)
    const version = namever.slice(separator + 1)
    const data = dependencies[name].find(package => package.version === version)

    return {
      count,
      data
    }
  })
  
  const top10 = result
    .sort((item1, item2) => item2.count - item1.count)
    .slice(0, 10)

  postMessage(top10)
}